New DISPLACEMENTS parser

alexmiame opened the pull request:
This PR moves the new DISPLACEMENTS file parser from the viperleed_jax repo to the main repository, since it makes more sense to have it here.
It does not (for now) use the new parser to actually read the DISPLACEMENTS file yet.

alexmiame committed:
feat: repr for constrain line

alexmiame committed:
doc,minor

alexmiame committed:
fix: update

alexmiame committed:
fix: empty sep

alexmiame committed:
fix: unpack targets

alexmiame committed:
fix: return value

alexmiame committed:
test, feat: tests for constrain line

alexmiame committed:
refactor: direct parsing in Line objects

alexmiame committed:
doc: improve doc for displacements reader class

alexmiame committed:
feat: logger

alexmiame committed:
feat,refactor!: remove regex module

Since it's been replaced by the more robust token and line parsers.

alexmiame committed:
fix: imports in token init

alexmiame committed:
fix!: restore DIRECTION_PATTERN

Was deleted with regex.py in 38b30942a2f99a01970ba406ef9d8fdfbbdd38e3

alexmiame committed:
todo: to discuss

alexmiame committed:
fix: logical error

linked assignments to single target make sense if target selects more than one atom

alexmiame committed:
test: change test data file

see 39c8f9835d92f1bde815cf5f8f7253ac0413ef44

alexmiame committed:
test,feat: update reader tests WIP

alexmiame committed:
test: update reader tests

alexmiame committed:
test,refactor: move fixture to conftest

alexmiame committed:
fix,minor: whitespace in message

alexmiame committed:
test,feat: update tests for displacements file class

alexmiame committed:
doc,minor

alexmiame committed:
refactor: move selection logic to atom basis

alexmiame committed:
fix,test: adapt tests for logic change in linked tag

alexmiame committed:
test,feat: test reading from file

alexmiame committed:
refactor: move perturbation type

alexmiame committed:
refactor,test: move test for perturbation type

alexmiame committed:
fix: import

alexmiame committed:
fix: import

alexmiame committed:
fix: import

alexmiame committed:
fix: import

alexmiame committed:
fix: repr of constraint line

alexmiame committed:
style,minor

alexmiame committed:
feat,test: zxy convention in direction token

alexmiame committed:
fix: target token edge cases

Fail on number or layer only target tokens

alexmiame committed:
test,feat: updated tests for invalid target tokens

alexmiame committed:
fix,test: target

alexmiame committed:
test: update lines tests

alexmiame committed:
Merge DISPLACEMENTS parser from viperleed_jax

alexmiame committed:
fix: import paths

alexmiame committed:
fix,test: path to files

alexmiame committed:
feat,chore: file header

alexmiame committed:
feat,doc: file headers

alexmiame committed:
Merge branch '0.14.1' into new-displacements-parser

alexmiame committed:
fix: rename to avoid name conflict

name conflict with old displacements.py

alexmiame committed:
fix: wrong rename

michele-riva committed:
Merge branch '0.14.1' into new-displacements-parser

michele-riva commented:
Sure, I will. Probably tonight or tomorrow morning.

alexmiame committed:
test,fix: invalid constrained DISPLACEMENTS

alexmiame committed:
test,fix: more invalid DISPLACEMENTS

alexmiame committed:
feat,test: use 3D displacements in test fiel

michele-riva commented on src/viperleed/calc/files/new_displacements/lines.py:
```diff
@@ -0,0 +1,504 @@
+        A tuple containing the targets string and the direction string. Either
+        string may be empty if the corresponding part was not found.
+    """
+    match = _DIR_AT_END.search(targets_and_direction)
```
Probably want to ensure there is only one of those.

michele-riva commented on src/viperleed/calc/files/new_displacements/reader.py:
```diff
@@ -0,0 +1,144 @@
+
+SEARCH_HEADER_PATTERN = re.compile(r'^=+\s+(?i:search)\s+(.*)$')
+SECTION_HEADER_PATTERN = re.compile(
+    r'^=+\s*(OFFSETS|GEO_DELTA|VIB_DELTA|OCC_DELTA|CONSTRAIN)$'
```
I believe the current syntax requires a single `=`. Of course we can make it more lenient.

michele-riva commented on src/viperleed/calc/files/new_displacements/reader.py:
```diff
@@ -0,0 +1,144 @@
+    VibDeltaLine,
+)
+
+SEARCH_HEADER_PATTERN = re.compile(r'^=+\s+(?i:search)\s+(.*)$')
```
I think the current syntax requires `==`. We can make it more lenient if there is no risk of confusing this with the sections.

michele-riva commented on src/viperleed/calc/files/new_displacements/reader.py:
```diff
@@ -0,0 +1,144 @@
+
+        self.current_section = None
+
+    def __next__(self):
+        """Read the next line from the file."""
+        for line in self._file_obj:
+            self._current_line += 1
+            line_content = self._read_one_line(line)
+            if line_content is not None:
+                return line_content
+        raise StopIteration
```
Do we really need this to override the base-class implementation? We could just raise `ShouldSkipLineError` in `_read_on_line` and let the base class do its job.

michele-riva commented on src/viperleed/calc/files/new_displacements/reader.py:
```diff
@@ -0,0 +1,144 @@
+        # check for search header
+        search_header_match = re.match(SEARCH_HEADER_PATTERN, line)
+        if search_header_match:
+            label = search_header_match.group(1)
```
Perhaps give the group a name. It's easier to read.

michele-riva commented on src/viperleed/calc/files/new_displacements/reader.py:
```diff
@@ -0,0 +1,144 @@
+        """Initialize instance."""
+        super().__init__(filename, noisy)
+
+        self.current_section = None
```
I would make this private not to risk accessing it an editing from outside.

michele-riva commented on src/viperleed/calc/files/new_displacements/reader.py:
```diff
@@ -0,0 +1,144 @@
+        """Get content from line."""
+        # Decide based on the current section
+        section = DisplacementFileSections[self.current_section]
+        if section is DisplacementFileSections.OFFSETS:
+            return OffsetsLine(line)
+        if section is DisplacementFileSections.GEO_DELTA:
+            return GeoDeltaLine(line)
+        if section is DisplacementFileSections.VIB_DELTA:
+            return VibDeltaLine(line)
+        if section is DisplacementFileSections.OCC_DELTA:
+            return OccDeltaLine(line)
+        if section is DisplacementFileSections.CONSTRAIN:
+            return ConstraintLine(line)
```
Perhaps more extensible with a `section_to_line_parser` dict? Or, better, set the Enum values directly to the line parsers rather than auto-assigning them. Then it would just be

```python
if not section:
    raise ...
parser = section.value
return parser(line)
```

michele-riva commented on src/viperleed/calc/files/new_displacements/reader.py:
```diff
@@ -0,0 +1,144 @@
+        section_header_match = re.match(SECTION_HEADER_PATTERN, line)
+        if section_header_match:
+            section = section_header_match.group(1)
+            self.current_section = section
```
Would perhaps be good to store directly the DisplacementsFileSection here instead of the string.

michele-riva commented on src/viperleed/calc/files/new_displacements/tl_backend.py:
```diff
@@ -0,0 +1,60 @@
+"""Module tl_backend of viperleed.files.displacements."""
```
(Part of) this module can probably be moved higher up the calc tree so we will be ready to support more backends, not only search-related. (See my suggestion of having a more generic `BACKEND` parameter in #333.)

michele-riva commented on src/viperleed/calc/files/new_displacements/tl_backend.py:
```diff
@@ -0,0 +1,60 @@
+__created__ = "2024-10-15"
+__license__ = "GPLv3+"
+
+class TensorLEEDBackend:
```
Probably an ABC? Would make it more extensible.

michele-riva commented on src/viperleed/calc/files/new_displacements/tl_backend.py:
```diff
@@ -0,0 +1,60 @@
+    if it can handle a certain type of search block.
+    """
+
+    def __init__(self, name, handle_search_block_func):
```
If an ABC, then you can make `replace_search_block` an `@abstractmethod` (can also be a classmethod so we can call it without even making an instance) and avoid passing a callable at creation. The `name` can be made into an abstract property.

michele-riva commented on src/viperleed/calc/files/new_displacements/tl_backend.py:
```diff
@@ -0,0 +1,60 @@
+def viplerleed_jax_search_block_handler_func(offsets_block, search_block):
+    """Handle the search block with the VIPERLEED backend."""
+    # no special handling needed, ViPErLEED jax can handle all search blocks
+    return
```
It should probably return a 1-tuple of the `search_block`?

michele-riva commented on src/viperleed/calc/files/new_displacements/tl_backend.py:
```diff
@@ -0,0 +1,60 @@
+
+        Parameters
+        ----------
+        offsets_block (OffsetsBlock): The offsets block.
```
I'm not sure I understand the need for the `offsets_block` here.

michele-riva reviewed (changes requested):
Looks very very good @amimre. A huge improvement on the current code. I have a few suggestions, and can then have a closer look at the code.

As you will see, one main logical bit seems missing: nested loops.

Of course, the other thing that would need to be implemented is the actual use of the `DisplacementsFile` object with a slab.

alexmiame committed:
style,minor

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
feat: suggestions

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

amimre commented on src/viperleed/calc/files/new_displacements/tokens/direction.py:
```diff
@@ -0,0 +1,152 @@
+    movements or constraints for geometric displacements. They can specify one,
+    two, or three dimensions of movements. The direction specifier can be given
+    as one of:
+    - 'xyz'
```
As I mentioned, I don't have the capacity to deal with the parsing of features not needed for the jax version atm, since there's plenty of other things that need doing before.
I don't intend for the new parser to replace the old one immediately anyhow.

alexmiame committed:
feat: more suggestions

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
doc: suggestion

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

amimre commented on src/viperleed/calc/files/new_displacements/tokens/range.py:
```diff
@@ -0,0 +1,101 @@
+                    f'step={self.step})')
+        return f'RangeToken(start={self.start}, stop={self.stop})'
+
+def _check_step(step):
+    """Check if the step is positive."""
+    if step is not None and step <= 0:
+        msg = f'Step must be positive: "{step}"'
+        raise RangeTokenParserError(msg)
```
We do?
Maybe we shouldn't since that seems somewhat confusing...

alexmiame committed:
feat: implement suggestions

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
feat: implement suggestions

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
doc: implement suggestions

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
style: abstract property

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
feat,doc: note and abc

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
feat: remove search block argument

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
feat: implement suggestion

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
feat,style: implement suggestions

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

amimre commented on src/viperleed/calc/files/new_displacements/file.py:
```diff
@@ -0,0 +1,247 @@
+        self.lines.append(line)
+
+
+class DisplacementsFile:
```
Do we currently properly support nested loops in the DISPLACEMENTS file?
And more importantly, do we want to continue doing so?

michele-riva replied:
I think we do support them, yes. But we'd need to add some tests to confirm that we indeed do. I think it is worth supporting them, yes. That's how we originally envisioned it. Perhaps @fkraushofer can comment further on this.

fkraushofer replied:
Yes, we do - and I think it's worth keeping for the TensErLEED backend. One usecase is an inner loop doing inplane and an outer loop doing z+vib optimization. This can make sense to ensure the inplane is actually converged, and because z+vib tends to be significantly more expensive (because it's a 2D-ish search).

With the new backend, I see no great need for loops in general - but as long as we have them, I think it's not particularly hard to support nesting as well.

fkraushofer replied:
Concerning implementation though, I think you may be overthinking it a bit. Correctly treating loops (including nested) only requires a list of the displacement blocks in order, and a list of indices (start, end) for each loop. This is how it's currently implemented. I can give you a quick rundown of how that works in person if you like, since I just recently edited it too.

michele-riva replied:
> With the new backend, I see no great need for loops in general - but as long as we have them, I think it's not particularly hard to support nesting as well.

Agreed. We would then just need to decide whether the new backend will "unroll" all loops to construct a larger parameter space, or if we should still keep.

amimre replied:
Well, you can't unroll a loop with an undetermined number of iterations, so you need to keep the look information in the object anyhow.
Either way, the parser & reader won't deal with the indices per se.

alexmiame committed:
fix,style: __str__ instead of __repr__

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
fix,test: test for __str__ instead of __repr__

alexmiame committed:
fix,test: use of __str__

alexmiame committed:
test,fix: use of __str__ instead of __repr__

alexmiame committed:
fix: logic

alexmiame committed:
feat: use __str__ instead of __repr__ for lines

alexmiame committed:
fix: double == in search block header

alexmiame committed:
feat,refactor: clean up errors as per suggestion

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

michele-riva commented on src/viperleed/calc/classes/perturbation_type.py:
```diff
@@ -0,0 +1,37 @@
+from enum import Enum
+
+
+class PerturbationTypeError(ValueError):
```
Sounds good.

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/type.py:
```diff
@@ -0,0 +1,39 @@
+"""Module for the <type> token in the DISPLACEMENTS file."""
```
Sounds good

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/base.py:
```diff
@@ -0,0 +1,23 @@
+
+from abc import ABC, abstractmethod
+
+class TokenParserError(ValueError):
```
Then perhaps we can make it be an `Exception` subclass rather than `ValueError`. We don't want to risk catching it in `except ValueError` blocks.

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/direction.py:
```diff
@@ -0,0 +1,152 @@
+        return f'DirectionToken(vectors={self.vectors}, dof={self.dof})'
+
+
+def _check_unsupported_directions(direction_str):
```
Sure. I'll un-resolve this so we keep this in mind.

michele-riva committed:
Merge branch '0.14.1' into new-displacements-parser

alexmiame committed:
Update target.py

Co-authored-by: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
refactor: suggestion

alexmiame committed:
fix: used DisplacementsSyntaxError

alexmiame committed:
fix,test: use __str__

alexmiame committed:
feat: classes for header lines

alexmiame committed:
fix: correct errors

alexmiame committed:
feat,fix: use section enum

alexmiame committed:
fix: errrors

alexmiame committed:
feat: rework reader implementation

alexmiame committed:
refactor: rename to PerturbationMode

alexmiame committed:
feat,refactor: rename files to perturbation_mode

alexmiame committed:
feat,refactor: rename TypeToken to ModeToken

alexmiame committed:
feat: better layer syntax

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
feat: range parsing via calc.lib.string_utils

Co-Authored-By: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
test: more layer syntax tests

alexmiame committed:
test: more lenient ranges

alexmiame committed:
feat: use base class __next__

alexmiame committed:
fix,test: wrong import

alexmiame committed:
fix,feat: use of sections dict

alexmiame committed:
test: offsets example file

alexmiame committed:
feat: change exception base class

alexmiame committed:
feat: vetors xyz & zxy support

alexmiame committed:
feat: direction token super class

alexmiame committed:
fix: error name

alexmiame committed:
test,fix: error class

alexmiame committed:
todo: unsupported directions

alexmiame committed:
style

alexmiame committed:
doc,minor

alexmiame committed:
feat: shared EPS for DISPLACEMENTS

alexmiame committed:
todo: step mandatory for TensErLEED

alexmiame committed:
feat: casting to square matrix

alexmiame committed:
feat: normalize brackets in direction token

alexmiame committed:
doc,fix

alexmiame committed:
fix: catch ValueError

alexmiame committed:
test: adapt tests for real cases

alexmiame committed:
style, minro

alexmiame committed:
fix,test: redundant match all regex

alexmiame committed:
feat: return NotImplemented in __eq__

alexmiame committed:
feat: make sure there is one match only

alexmiame committed:
feat: allow negative step

alexmiame committed:
fix: error message typo

alexmiame committed:
feat: enable 1-3 dimensional offsets

alexmiame committed:
test: adapt offset tests

alexmiame committed:
test: adapt tests for negative steps

fkraushofer commented on src/viperleed/calc/files/new_displacements/lines.py:
```diff
@@ -0,0 +1,504 @@
+    """Class to parse lines in the OCC_DELTA block of DISPLACEMENTS.
+
+    Lines in the OCC_DELTA block are of the form:
+        <target> [, <target>] = <element> <range> [, <element> <range> ...]
```
I'm happy to deprecate this one. Was never a big fan, and I can't imagine anyone is using it since it's still very complicated to understand the syntax, and as you say, the same can be accomplished by constraints.

alexmiame committed:
feat: deprecated combined element ranges

alexmiame committed:
chore: unused import

michele-riva committed:
Merge branch '0.14.1' into new-displacements-parser

michele-riva commented on src/viperleed/calc/files/new_displacements/lines.py:
```diff
@@ -0,0 +1,504 @@
+        # check if the last part is a direction
+        targets_str, dir_str = separate_direction_from_targets(self._lhs)
+        if dir_str:
+            raise InvalidDisplacementsSyntaxError(self.invalid_format_msg)
```
Sorry, I screwed this up. I did not see that the classes had an `.invalid_format_msg` attribute, so I thought that this line would raise an `AttributeError` when executed. I think `DisplacementsSyntaxError` here is fine.

alexmiame committed:
Merge branch 'new-displacements-parser' of https://github.com/viperleed/viperleed into new-displacements-parser

alexmiame committed:
feat: displacements segment base class

alexmiame committed:
feat: imports

alexmiame committed:
feat: segments

alexmiame committed:
feat: error

alexmiame committed:
refactor: move classes

alexmiame committed:
refactor,fix: loop block

alexmiame committed:
chore: redundant

alexmiame committed:
test: preliminary tests

alexmiame committed:
feat: validated block

alexmiame committed:
feat: more safety with loop

alexmiame committed:
chore: remove debug print

alexmiame committed:
feat,refactor: displacements file

alexmiame committed:
feat: imports

alexmiame committed:
fix: missing tab

alexmiame committed:
feat: parsing

alexmiame committed:
feat: file validation

alexmiame committed:
test,feat: all DISPLACEMENTS fixture

alexmiame committed:
test, feat: update file reading tests

alexmiame committed:
fix,refactor: offsets block handling

alexmiame committed:
feat: use invalid_format_msg

alexmiame committed:
style: whitespace

alexmiame committed:
feat: rename offsets

alexmiame committed:
feat: iterator class WIP

alexmiame committed:
feat,refactor: move interation to flie

alexmiame committed:
feat: iteration in loop block

alexmiame committed:
doc: docstrings

amimre commented:
Alright, I worked through the issues raised so far.
I really don't have the capacity to work on this any more for the time being, except as far as it is required for the viperleed-jax plugin and the parameter tree, as I really need to work on finishing those for the time being.

alexmiame committed:
feat,style: check has been read

alexmiame committed:
test: adapt has been read

alexmiame committed:
feat,fix: access to search block lines

alexmiame committed:
style: whitespace

alexmiame committed:
feat: remove non-useful log message

alexmiame committed:
feat: add anytree dependency

alexmiame committed:
feat: total occupation token

alexmiame committed:
doc: total occ docstring

alexmiame committed:
test: tests for total occupation token

alexmiame committed:
chore: imports

alexmiame committed:
feat: use total occ token

michele-riva committed:
Merge branch '0.14.1' into new-displacements-parser

michele-riva commented:
I can have a look at it

michele-riva commented on tests/calc/files/displacements/test_lines.py:
```diff
@@ -0,0 +1,454 @@
+                TypeToken('geo'),
+                [TargetToken('A')],
+                TargetToken('B'),
+                LinearOperationToken('[1 0 0]'),
```
`LinearOperationTokenParserError: Invalid linear operation format: "[1 0 0]". Expected a 1x1, 2x2, or 3x3 matrix.`

michele-riva commented on tests/calc/files/displacements/test_lines.py:
```diff
@@ -0,0 +1,454 @@
+        # Type
+        assert con.type == exp_type
+        # Targets
+        if 'linked' in line:
+            for target, exp_target in zip(con.targets[:-1], exp_targets):
+                assert target == exp_target
+        else:
+            for target, exp_target in zip(con.targets, exp_targets):
+                assert target == exp_target
```
```suggestion
        act_targets = con.targets
        if 'linked' in line:
            act_targets = act_targets[:-1]
        assert len(act_targets) == len(exp_targets)
        for target, exp_target in zip(act_targets, exp_targets):
            assert target == exp_target
```

Also, add a comment as to why `[:-1]`.

michele-riva commented on tests/calc/files/displacements/test_reader.py:
```diff
@@ -0,0 +1,33 @@
+"""Tests for files/files/displacements/reader.py."""
+
+__authors__ = ('Alexander M. Imre (@amimre)',)
```
Some missing dunders. Check also all other test files. Some are missing all of them.

michele-riva commented on tests/calc/files/displacements/test_reader.py:
```diff
@@ -0,0 +1,33 @@
+    with DisplacementsReader(path) as reader:
+        parsed_lines = list(reader)
+    assert len(parsed_lines) == len(expected_lines)
+    for parsed, expected in zip(parsed_lines, expected_lines):
```
Maybe you want to unpack this `expected` already when iterating into an `(exp_cls, exp_attrs)` tuple?

michele-riva commented on tests/calc/files/displacements/test_segments.py:
```diff
@@ -0,0 +1,59 @@
+    SectionHeaderLine,
+    VibDeltaLine,
+)
+from viperleed.calc.files.new_displacements.segments import SearchBlock, LoopBlock
```
```suggestion
from viperleed.calc.files.new_displacements.segments import (
    LoopBlock,
    SearchBlock,
    )
```

michele-riva commented on tests/calc/files/displacements/test_segments.py:
```diff
@@ -0,0 +1,59 @@
+            LoopMarkerLine("</loop>"),
+        ]
+    )
+    loop_block = LoopBlock(header)
```
Shouldn't this complain that its only child is another `LoopBlock`?

michele-riva commented on tests/calc/files/displacements/tokens/test_element.py:
```diff
@@ -0,0 +1,41 @@
+__authors__ = ('Alexander M. Imre (@amimre)',)
+__created__ = '2025-05-13'
+
+# test_element_token.py
```
```suggestion
```

michele-riva commented on tests/calc/files/displacements/tokens/test_offset.py:
```diff
@@ -0,0 +1,54 @@
+import pytest
```
Missing module docstring. Check also other tests, and conform to our standard (see any of the calc tests modules)

michele-riva commented on tests/calc/files/displacements/tokens/test_offset.py:
```diff
@@ -0,0 +1,54 @@
+    assert str(token) == f'OffsetToken(offset={token.offset})'
+
+
+def test_init_invalid_count():
+    for bad in ['', '1 2 3 4', 'a b c']:
+        with pytest.raises(OffsetTokenParserError) as exc:
+            OffsetToken(bad)
```
`@parametrize`; `exc` unused.

michele-riva commented on tests/calc/files/displacements/tokens/test_offset.py:
```diff
@@ -0,0 +1,54 @@
+            OffsetToken(bad)
+
+
+def test_init_non_numeric():
+    for bad in ['foo', '1.0x', '--3']:
+        with pytest.raises(OffsetTokenParserError) as exc:
+            OffsetToken(bad)
+        assert 'Non-numeric value' in str(exc.value)
```
`@parametrize`

michele-riva commented on tests/calc/files/displacements/tokens/test_range.py:
```diff
@@ -0,0 +1,112 @@
+import pytest
+
+from viperleed.calc.files.new_displacements.tokens.range import (
+    RangeToken, RangeTokenParserError
```
```suggestion
    RangeToken,
    RangeTokenParserError,
```

michele-riva commented on tests/calc/files/displacements/tokens/test_range.py:
```diff
@@ -0,0 +1,112 @@
+    assert str(dr) == 'RangeToken(start=1.0, stop=2.0)'
+
+    drs = RangeToken('1 2 0.25')
+    assert str(drs) == 'RangeToken(start=1.0, stop=2.0, step=0.25)'
```
Missing newline at end

michele-riva commented on tests/calc/files/displacements/tokens/test_target.py:
```diff
@@ -0,0 +1,126 @@
+    'raw',
+    [
+        # combined layer and number
+        'A L(1-2) 3',
```
Is there a reason why we shouldn't support this?

amimre replied:
What would be the point?
If I say layers x-y, and atom z it's unclear what the user means.
If atom z is in these layers, does it mean only the atom or everyting? What if the atom is not in those layers?

michele-riva commented on tests/calc/files/displacements/tokens/test_target.py:
```diff
@@ -0,0 +1,126 @@
+        # wrongly formatted layer
+    ])
+def test_invalid_target_raises(raw):
+    with pytest.raises(TargetingError) as exc:
```
exc unused

michele-riva commented on tests/calc/files/displacements/tokens/test_target.py:
```diff
@@ -0,0 +1,126 @@
+
+
+def test_empty_string_raises():
+    with pytest.raises(TargetingError) as exc:
```
You can use the `match` keyword of `pytest.raises` rather than checking manually

michele-riva commented on tests/calc/files/displacements/tokens/test_target.py:
```diff
@@ -0,0 +1,126 @@
+    assert a == b
+    # different label
+    d = TargetToken('B 1-2')
+    assert not (a == d)
```
```suggestion
    assert a != d
```

michele-riva replied:
same below

michele-riva commented on tests/calc/files/displacements/tokens/test_target.py:
```diff
@@ -0,0 +1,126 @@
+    # Should match XANY, XfooYbar, etc.
+    assert r.match('XANY')
+    assert r.match('XfooYbar')
+    assert not r.match('XYX') == None  # ensure it matches "XYX"
```
```suggestion
    assert r.match('XYX')
```

michele-riva reviewed (changes requested):
Good job @amimre. Mostly small stuff to change.

alexmiame committed:
Merge branch 'new-displacements-parser' of https://github.com/viperleed/viperleed into new-displacements-parser

alexmiame committed:
Apply suggestions from code review

Co-authored-by: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
style: apply comments

alexmiame committed:
refactor: move displacement eps constant

alexmiame committed:
style,refactor: basic formatting

alexmiame committed:
Merge branch 'search-job-class' into new-displacements-parser

alexmiame committed:
style: more autoformatting

alexmiame committed:
refactor: more formatting

alexmiame committed:
Merge branch 'search-job-class' into new-displacements-parser

alexmiame committed:
style,refactor: more changes

alexmiame committed:
style,refactor: several more comments

alexmiame committed:
fix: missing Nones

alexmiame committed:
style: various comments by Michele

alexmiame committed:
chore,refactor: many comments by Michele

alexmiame committed:
feat: add ruff setting

alexmiame committed:
fix: loop handling

alexmiame committed:
fix,test: add more tests for loop blocks

alexmiame committed:
fix,feat: case insensitve perturbation mode

alexmiame committed:
feat,refactor: more comments

alexmiame committed:
feat: various string formatings

alexmiame committed:
fix: renamed type to mode

alexmiame committed:
refactor: backends

alexmiame committed:
feat: strings for parsed lines

alexmiame committed:
doc: fixed documentation

alexmiame committed:
fix: use eps

alexmiame committed:
fix: formatting of rhs for occ line

alexmiame committed:
fix: redefinition

alexmiame committed:
fix: check offset DOF

alexmiame committed:
fix: overly strict check

alexmiame committed:
style: no type hint

alexmiame committed:
test, fix: remove invalid test

alexmiame committed:
test,fix: mark xfail

alexmiame committed:
refactor: rename test file directory

alexmiame committed:
test,feat: test without search block header

add a simple DISPLACMENETS file that has no header

michele-riva committed:
Merge branch '0.14.1' into new-displacements-parser

michele-riva reviewed (changes requested):
Sorry @amimre, some more suggestions, but I think we're converging.

michele-riva reviewed (changes requested):
Sorry @amimre, some more suggestions, but I think we're converging.

alexmiame committed:
chore,style,minor: file headers and doc typos

alexmiame committed:
chore: remove redundant redefinition

alexmiame committed:
fix: remove double import

alexmiame committed:
style,minor

alexmiame committed:
chore: move files back

alexmiame committed:
doc,style,minor: file headers and type hints

alexmiame committed:
minro

alexmiame committed:
refactor: remove unneeded init

alexmiame committed:
style,minor

alexmiame committed:
fix: add super init

(though probably not necessary)

alexmiame committed:
doc,fix

alexmiame committed:
refactor: detect inconsitent OFFSETS blocks

alexmiame committed:
refactor: segments

alexmiame committed:
doc,fix

alexmiame committed:
fix: validate direction string

alexmiame committed:
feat: close check for 2 and 3 D

fkraushofer commented on src/viperleed/calc/files/new_displacements/tokens/element.py:
```diff
@@ -0,0 +1,54 @@
+    chemical elements). It is used in the OCC_DELTA block and in lines relating
+    to chemical parameters in the OFFSETS and CONSTRAIN blocks.
+    The <element> token is specified by the user as a simple (case-insensitive)
+    string corresponding to the Symbol of a element from the periodic table.
```
Hmm, I think both functionalities are pretty much unused at the moment, and probably noone would miss them if they were gone. 

I can see the use in explicitly defining vacancies, especially if you wanted to limit them to some range - say, you mix 2 elements in arbitrary amounts (which I guess would by default also allow 100% vacancies?), and you want to also set the restriction that at most 10% are vacancies. Sure, you could do that via constraints, but actually, doing it here by saying 0-10% vacancies would be quite intuitive. But also, depending on how it's implemented, I can see that this might be a substantial amount of work... In that case, I'd be fine with making a nice-to-have issue to come back to this later (maybe), and leave it for now.

Defining some element as constant as in the Ti example I find actively confusing and inconsistent with the rest of DISPLACEMENTS. Strongly in favour of deprecating.

michele-riva replied:
Fine with me to deprecate the no-stop case of Ti.

@amimre: concerning bounds for vacancies, I believe it would be rather straightforward to implement in JAX: It would probably be just a matter of using the user-defined bounds rather than calculating them out of those of the other elements.

amimre replied:
Well, I'd say semi-straightforward – but still work to implement bounds, parsing, tests etc.. I'll table it for now.

alexmiame committed:
doc: class headers

alexmiame committed:
refactor init subclass

alexmiame committed:
refactor

alexmiame committed:
style,minor

alexmiame committed:
fix,minor

alexmiame committed:
feat,fix: import file class in init

alexmiame committed:
fix: broken header line logic

alexmiame committed:
fix,refactor: don't use .type

alexmiame committed:
fix: mode token use

alexmiame committed:
Merge branch '0.14.1' into new-displacements-parser

alexmiame committed:
feat: apply suggestions

alexmiame committed:
feat,minor: make into class attribute

alexmiame committed:
feat, refactor: init_subclass

alexmiame committed:
doc,minor

alexmiame committed:
feat: check subsegment type

alexmiame committed:
feat: simplified logic

alexmiame committed:
doc: add method docstring

alexmiame committed:
doc: fix incorrect type

alexmiame commented on src/viperleed/calc/files/new_displacements/tokens/target.py:
```diff
@@ -0,0 +1,108 @@
+    """Class to handle the <target> token in the DISPLACEMENTS file."""
+
+    def __init__(self, target_str):
+        self.target_str = target_str.strip()
```
Strings are parsed at init for other tokens as well.

alexmiame committed:
feat: use tuple instead of list

alexmiame committed:
feat,fix: target token comparison

alexmiame committed:
doc,todo: add todo note

alexmiame committed:
minor,test: suggested change

alexmiame committed:
test,fix: tuple instead of list

alexmiame committed:
test,feat: create id from filename

alexmiame committed:
test,fix: offset token has no mode

alexmiame committed:
test,feat: tests for 2D and 3D offest tokens

alexmiame committed:
test,feat: split supported and unsupported files

alexmiame committed:
test,minor: whitespace

alexmiame committed:
fix: require whitespace before direction token

Otherwise a LHS of e.g. "Pt_sub" would be interpreted as target "Pt_su" plus direction "b".

alexmiame committed:
doc: docstring line lengths

alexmiame committed:
test, feat: test next method for file

alexmiame committed:
test,fix: remove nonsensical direction only case

alexmiame committed:
fix: strip RHS for total tag in total occ constraint

alexmiame committed:
test,feat: add test for total occ constraint

alexmiame committed:
Merge branch '0.14.2' into new-displacements-parser

alexmiame committed:
fix: alex

alexmiame committed:
doc,fix: missing dunders

alexmiame committed:
doc,fix: extend copyright to 2026

michele-riva commented on src/viperleed/calc/classes/perturbation_mode.py:
```diff
@@ -0,0 +1,34 @@
+        return self.value
+
+    @classmethod
+    def from_string(cls, mode_string: str):
+        """Convert a string to a PerturbationType enum."""
```
```suggestion
    def from_string(cls, mode_string):
        """Convert a string to a PerturbationMode enum."""
```

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/__init__.py:
```diff
@@ -0,0 +1,16 @@
+__created__ = '2025-05-12'
+__license__ = 'GPLv3+'
+
+from .base import TokenParserError
+from .direction import CartesianDirectionToken
+from .element import ElementToken
+from .linear_operation import LinearOperationToken
+from .mode import ModeToken
+from .offset import OffsetToken
+from .range_ import RangeToken
+from .target import TargetToken
+from .total_occupation import TotalOccupationToken
```
Do we need all these init imports? May lead to circular imports later on

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/direction.py:
```diff
@@ -0,0 +1,205 @@
+    dof : int
+        The number of degrees of freedom (DOF), either 1, 2, or 3.
+    vectors_xyz : ndarray, shape (dof, 3)
+        A tuple of one, two or three orthonormal vectors spanning the
```
perhaps here "tuple" is not the correct word, as it is an ndarray

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/direction.py:
```diff
@@ -0,0 +1,205 @@
+        vecs = [self._get_basis_vector(c) for c in direction_str]
+        return np.array(vecs), len(vecs)
+
+    def _get_basis_vector(self, label):
```
the three private methods here don't use self and could be `@staticmethod`s

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/direction.py:
```diff
@@ -0,0 +1,205 @@
+            raise ValueError(msg)
+        return vec / norm
+
+    def __eq__(self, other):
```
These dunders are public, and should appear before private methods

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/linear_operation.py:
```diff
@@ -0,0 +1,96 @@
+            cleaned = re.sub(r'(?<=\d)\s+(?=\d)', ', ', cleaned)
+            # add commas between brackets
+            cleaned = re.sub(r'\]\s+\[', '], [', cleaned)
+        # replace round brackets with square brackets
+        cleaned = cleaned.replace('(', '[').replace(')', ']')
```
Perhaps this should be done before the comma sanitization? Otherwise there would be no commas added between two round brackets if it is missing

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/linear_operation.py:
```diff
@@ -0,0 +1,96 @@
+        self.arr = arr
+
+    @classmethod
+    def from_array(cls, arr) -> 'LinearOperationToken':
```
```suggestion
    def from_array(cls, arr):
```

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/linear_operation.py:
```diff
@@ -0,0 +1,96 @@
+    def from_array(cls, arr) -> 'LinearOperationToken':
+        """Alternate constructor using numeric values directly."""
+        inst = cls.__new__(cls)
+        inst.arr = np.array(arr, dtype=float)
```
This implementation skips all the validation on arr compared to the string version. Perhaps the validation of the array should be factored out from init and reused here too?

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/offset.py:
```diff
@@ -0,0 +1,82 @@
+            raise OffsetTokenParserError(msg)
+
+        try:
+            offset_values = np.array([float(p) for p in parts], dtype=float)
```
I think we can skip the conversion to a list of float here, and let numpy take care of it:

```suggestion
            offset_values = np.array(parts, dtype=float)
```

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/element.py:
```diff
@@ -0,0 +1,56 @@
+        The string containing the element to be parsed.
+    """
+
+    def __init__(self, element_str: str):
```
```suggestion
    def __init__(self, element_str):
```

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/offset.py:
```diff
@@ -0,0 +1,82 @@
+        If parsing fails.
+    """
+
+    def __init__(self, offset_str: str):
```
```suggestion
    def __init__(self, offset_str):
```

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/offset.py:
```diff
@@ -0,0 +1,82 @@
+        return self.offset.size
+
+    @classmethod
+    def from_floats(cls, *offsets: float) -> 'OffsetToken':
```
```suggestion
    def from_floats(cls, *offsets):
```

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/offset.py:
```diff
@@ -0,0 +1,82 @@
+    @classmethod
+    def from_floats(cls, *offsets: float) -> 'OffsetToken':
+        """Alternate constructor using 1–3 numeric values directly."""
+        if not (1 <= len(offsets) <= 3):
```
The validation code in this method is very similar to that in init. Consider refactoring to a _validate_offsets staticmethod

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/range_.py:
```diff
@@ -0,0 +1,92 @@
+        return self.step is not None
+
+    @classmethod
+    def from_floats(cls, start, stop, step=None) -> 'RangeToken':
```
```suggestion
    def from_floats(cls, start, stop, step=None):
```

michele-riva commented on src/viperleed/calc/files/new_displacements/tokens/total_occupation.py:
```diff
@@ -0,0 +1,64 @@
+        The string containing the total occupation to be parsed.
+    """
+
+    def __init__(self, total_occupation_str: str):
```
```suggestion
    def __init__(self, total_occupation_str):
```

michele-riva commented on src/viperleed/calc/files/new_displacements/__init__.py:
```diff
@@ -0,0 +1,8 @@
+__created__ = '2024-10-04'
+__license__ = 'GPLv3+'
+
+from .file import DisplacementsFile
```
Same comment as for other package init files

michele-riva commented on src/viperleed/calc/files/new_displacements/file.py:
```diff
@@ -0,0 +1,213 @@
+        super().__init__()
+        self.name = self._render_name
+        self._has_been_read = False
+        # an OFFSETS block is only allowed at the very beginning of the file
+        # we check this by setting this flag to True after the first block
```
Is this comment still relevant? _has_been_read is set to True as soon as we start reading, not after the first block. Perhaps this referred to an earlier implementation?

michele-riva commented on src/viperleed/calc/files/new_displacements/file.py:
```diff
@@ -0,0 +1,213 @@
+        if not self.children:
+            raise DisplacementsSyntaxError('The file is empty.')
+
+        # check if there is more than one OFFSETS block
+        if (
+            len(
+                [
+                    block
+                    for block in self.descendants
+                    if isinstance(block, OffsetsBlock)
+                ]
+            )
+            > 1
+        ):
+            raise OffsetsNotAtBeginningError(
+                'There can only be one OFFSETS block at the beginning of the '
+                'DISPLACEMENTS file.'
+            )
```
This bit seems redundant with the rest (though the comment is probably OK while the next one is misplaced)

michele-riva commented on src/viperleed/calc/files/new_displacements/segments.py:
```diff
@@ -0,0 +1,367 @@
+        """Return the lines belonging to this segment."""
+        return tuple(self._lines)
+
+    def read_lines(self, lines):
```
Missing docstring

michele-riva commented on src/viperleed/calc/files/new_displacements/segments.py:
```diff
@@ -0,0 +1,367 @@
+
+    @abstractmethod
+    def _belongs_to_me(self, line):
+        """Return whether `line` is a header for this segment."""
```
This docstring does not seem correct

michele-riva commented on src/viperleed/calc/files/new_displacements/segments.py:
```diff
@@ -0,0 +1,367 @@
+    def _belongs_to_me(self, line):
+        """Loop end lines belong to this block."""
+        # is the loop already closed?
+        if self.lines and self.lines[-1].kind == 'end':
```
Probably should check that the last line is a LoopMarkerLine too, otherwise AttributeError on accessing .kind. Something like this would work:

```suggestion
        last_line = self.lines[-1] if self.lines else None
        if isinstance(last_line, LoopMarkerLine) and last_line.kind == 'end':
```

michele-riva commented on src/viperleed/calc/files/new_displacements/lines.py:
```diff
@@ -0,0 +1,694 @@
+LOOP_END_PATTERN = re.compile(r'<\\loop>|</loop>', re.IGNORECASE)
+
+
+_BELOW_DEBUG = 2
```
Better to replace this with LOG_VERY_VERBOSE from calc.constants

michele-riva commented on src/viperleed/calc/files/new_displacements/lines.py:
```diff
@@ -0,0 +1,694 @@
+        = OFFSETS
+    """
+
+    section = 'OFFSETS'
```
This class attribute seems unused in the implementation (e.g., in `__str__`)

michele-riva commented on src/viperleed/calc/files/new_displacements/lines.py:
```diff
@@ -0,0 +1,694 @@
+    where <section> is one of the defined sections in the DISPLACEMENTS file.
+    """
+
+    def __init__(self, section: str):
```
```suggestion
    def __init__(self, section):
```

Remove also other type annotations below (and check also other modules)

michele-riva commented on src/viperleed/calc/files/new_displacements/lines.py:
```diff
@@ -0,0 +1,694 @@
+
+    def _treat_rhs(self):
+        # check for deprecated 'offset' tag
+        if 'offset' in self._rhs.lower():
```
I don't think that `offset` was ever supported in [CONSTRAIN blocks](https://www.viperleed.org/stable/content/calc/files/input/displacements/searchconstraints.html#searchconstraints). It (was) available as a left-hand-side flag for [GEO_DELTA](https://www.viperleed.org/stable/content/calc/files/input/displacements/geodelta.html#geodelta) and [VIB_DELTA](https://www.viperleed.org/stable/content/calc/files/input/displacements/vibdelta.html#vibdelta).

michele-riva commented on src/viperleed/calc/files/new_displacements/lines.py:
```diff
@@ -0,0 +1,694 @@
+            # parse the total occupation value
+            total_occupation_str = self._rhs[len('total ') :].strip()
+
+            # this is a special case, where the linear operation is
```
comment seems incomplete

michele-riva commented on src/viperleed/calc/files/new_displacements/lines.py:
```diff
@@ -0,0 +1,694 @@
+
+    Lines in the OFFSETS block are of the form:
+        <type> <target> [, <target> ...] [<direction>] = <offset>
+    where <target>, ... are tokens that are parsed by the OffsetToken class.
```
This does not seem correct

michele-riva committed:
chore: undo useless Ruff edits

Concern files that are unrelated to this PR

michele-riva committed:
chore: more useless Ruff edits

michele-riva committed:
chore: restore init.py

Was inadvertently removed

michele-riva committed:
chore: move back test file

Ended up in the wrong folder

michele-riva commented on tests/calc/files/new_displacements/tokens/test_target.py:
```diff
@@ -0,0 +1,133 @@
+    assert not regex.match('B')
+
+
+def test_invalid_number_format_raises():
+    with pytest.raises(TargetingError):
+        # non-integer in numeric list
+        TargetToken('A one two')
+    with pytest.raises(TargetingError):
+        # malformed layer spec
+        TargetToken('A L()')
```
Can be done in test_invalid_target_raises by adding these strings to the parametrization there

michele-riva commented on tests/calc/files/new_displacements/test_reader.py:
```diff
@@ -0,0 +1,33 @@
+
+def test_displacements_reader(mock_displacements_path_and_lines):
+    path, expected_lines = mock_displacements_path_and_lines
+    _compare_lines(path, expected_lines)
```
One fails with `'ConstraintLine' object has no attribute 'type'`

michele-riva commented on tests/calc/files/new_displacements/test_segments.py:
```diff
@@ -0,0 +1,106 @@
+    assert len(returned_lines) == 0
+
+
+def test_layered_loop_block():
```
Fails with `DisplacementsSyntaxError: Loop block cannot contain only another loop block.` at the read_lines call below

michele-riva reviewed (changes requested):
Some more suggestions on this one, but we're converging.
